### PR TITLE
Support for default table properties (CSS: `.ck-content`)

### LINKED ISSUE #8502: No way to set default alignment for images or tables.
When you insert images / tables there's no way to pick the default alignment. 

I looked at the alignment for images and there is a 'styles' which has an isDefault property but that's not actually working.

I also can't find any documentation on how to do this with tables either.

### LINKED ISSUE #9219: Default table properties via configuration
## 📝 Provide a description of the new feature

_What is the expected behavior of the proposed feature?_

e.g. How do we configure the default table styles? I want a 1px border with certain colors to be pre-filled.

Example:

```js
{
  table: {
    tableProperties: {
      defaults: {
        borderColor: 'red',
        borderWidth: '1',
        borderStyle: 'solid'
      }
   }
}
```

Upon the user inserting a table, they should see:

image

History: To avoid confusion, this remakes #6363's [original post](https://github.com/ckeditor/ckeditor5/issues/6363#issue-573975234) as its topic changed and it was closed. The new topic at #6841 is speaking about something different than setting default (clearing the value vs setting `none` to negate?)

---

If you'd like to see this feature implemented, add a 👍 reaction to this post.

### LINKED ISSUE #9536: upcastStyleToAttribute() helper is called in wrong context
## 📝 Provide detailed reproduction steps (if any)

https://github.com/ckeditor/ckeditor5/blob/2eb5c73b0f928f45c7b53e3b56b92539854713cd/packages/ckeditor5-table/src/converters/tableproperties.js#L10-L37

Replace the `model.value` function with the snippet:

```js
viewElement => {
	const normalized = viewElement.getNormalizedStyle( styleName );

	console.log( {
		viewElementName: viewElement.name,
		modelElementToSet: modelElement,
		modelAttributeToSet: styleName,
		attributeValue: reduceBoxSides ? reduceBoxSidesValue( normalized ) : normalized
	} );

	return reduceBoxSides ? reduceBoxSidesValue( normalized ) : normalized;
}
```

Then, open the [table properties manual test](http://localhost:8125/ckeditor5-table/tests/manual/tableproperties.html) and set the editor data with the following table:

```js
editor.setData(`<table style="float: left;width:300px;height:250px;border: dashed 3px hsl(0, 0%, 60%);background: #00f"> <tr> <td>Foo.</td> <td style="width:100px">Foo.</td> </tr> <tr> <td>Foo.</td> <td>Foo.</td> </tr> </table>`);
```

```html
	<table style="float: left;width:300px;height:250px;border: dashed 3px hsl(0, 0%, 60%);background: #00f">
		<tr>
			<td>Foo.</td>
			<td style="width:100px">Foo.</td>
		</tr>
		<tr>
			<td>Foo.</td>
			<td>Foo.</td>
		</tr>
	</table>
```

The helper function is used for upcasting all table and table cell properties. I'd like to focus on a single property, e.g. "width".

### ✔️ Expected result

I'd expect two logs for the `width` property.

1. ```{viewElementName: "td", modelElementToSet: "tableCell", modelAttributeToSet: "width", attributeValue: "100px"}```
2. ```{viewElementName: "table", modelElementToSet: "table", modelAttributeToSet: "width", attributeValue: "300px"}```

These logs are displayed in proper context: when upcasting the `table` (view) element, set the `table` (model) element. When upcasting the `td` (view) element, set the the `tableCell` (model) element.

Combinations (view->model): `td->table` and `table->tableCell` are invalid and should not be logged.

#### "width" property

```js
{viewElementName: "td", modelElementToSet: "table", modelAttributeToSet: "width", attributeValue: "100px"}
{viewElementName: "td", modelElementToSet: "tableCell", modelAttributeToSet: "width", attributeValue: "100px"}
{viewElementName: "table", modelElementToSet: "table", modelAttributeToSet: "width", attributeValue: "300px"}
{viewElementName: "table", modelElementToSet: "tableCell", modelAttributeToSet: "width", attributeValue: "300px"}
```

#### Full log

```js
{viewElementName: "td", modelElementToSet: "table", modelAttributeToSet: "width", attributeValue: "100px"}
{viewElementName: "td", modelElementToSet: "tableCell", modelAttributeToSet: "width", attributeValue: "100px"}
{viewElementName: "table", modelElementToSet: "table", modelAttributeToSet: "width", attributeValue: "300px"}
{viewElementName: "table", modelElementToSet: "table", modelAttributeToSet: "height", attributeValue: "250px"}
{viewElementName: "table", modelElementToSet: "table", modelAttributeToSet: "background-color", attributeValue: "#00f"}
{viewElementName: "table", modelElementToSet: "tableCell", modelAttributeToSet: "width", attributeValue: "300px"}
{viewElementName: "table", modelElementToSet: "tableCell", modelAttributeToSet: "height", attributeValue: "250px"}
{viewElementName: "table", modelElementToSet: "tableCell", modelAttributeToSet: "background-color", attributeValue: "#00f"}
```

## 📃 Other details

While working on the default table properties, I wanted to disallow upcasting properties if their values are equal to default properties specified in the editor configuration.

So, I passed the default value for each property and updated the `console.log`:

```js
console.log( {
	viewElementName: viewElement.name,
	modelElementToSet: modelElement,
	modelAttributeToSet: styleName,
	attributeValue: reduceBoxSides ? reduceBoxSidesValue( normalized ) : normalized,
	shouldSetValue: value !== defaultValue
} );
```

Input data:

```html
<table style="float: left;width:300px;height:250px;border: dashed 3px hsl(0, 0%, 60%);background: #00f">
	<tr>
		<td>Foo.</td>
	</tr>
</table>
```

Just a single table cell. And filtred console.logs calls (only `width` property is displayed):

```js
{viewElementName: "table", modelElementToSet: "table", modelAttributeToSet: "width", attributeValue: "300px", shouldSetValue: false}
{viewElementName: "table", modelElementToSet: "tableCell", modelAttributeToSet: "width", attributeValue: "300px", shouldSetValue: true}
```

The main question is – why the converter was called for the `tableCell` element if the input data does not contain any "style" attribute?

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Feature (table): Support for the default table properties. Read more about [the feature in the documentation](https://ckeditor.com/docs/ckeditor5/latest/features/table.html). Closes #8502. Closes #9219.

Fix (engine): The conversion upcast `elementToAttribute()` and `attributeToAttribute()` functions should not call the `model.value()` callback if the element will not be converted. Closes #9536.

MINOR BREAKING CHANGE (table): Clases `TableAlignmentCommand`, `TableBackgroundColorCommand`, `TableBorderColorCommand`, `TableBorderStyleCommand`, `TableBorderWidthCommand`, `TableHeightCommand`, `TablePropertyCommand`, `TableWidthCommand` requires the second argument called `defaultValue` which is the default value for the command.

MINOR BREAKING CHANGE (table): The `TablePropertiesView` class requires additional property in the object as the second constructor argument (`options.defaultTableProperties`).

---

### Additional information

🧐

### CODE REVIEW

File: packages/ckeditor5-table/src/tableproperties/tablepropertiesui.js [hunk 1]
```diff
@@ -81,6 +82,16 @@ export default class TablePropertiesUI extends Plugin {
...
 		const t = editor.t;
 
+		/**
+		 * The default table properties.
+		 *
+		 * @protected
+		 * @member {module:table/tableproperties~TablePropertiesOptions}
+		 */
+		this._defaultTableProperties = getNormalizedDefaultProperties( editor.config.get( 'table.tableProperties.defaultProperties' ), {
+			includeAlignmentProperty: true
```
Comment (niegowski) [inline packages/ckeditor5-table/src/tableproperties/tablepropertiesui.js:92]: This option seems to be not needed - it's the same in all calls.

File: packages/ckeditor5-table/docs/features/table.md [hunk 2]
```diff
@@ -205,6 +205,58 @@ ClassicEditor
 	.catch( ... );
 ```
 
+### Default table styles
+
+The default tables feature allows configuring the default look of tables in the editor. The configuration object should be synchronized with the {@link builds/guides/integration/content-styles editor content styles}.
```
Comment (niegowski) [inline packages/ckeditor5-table/docs/features/table.md:210]: The beginning of this sentence seems to miss sth.

Comment (niegowski) [inline packages/ckeditor5-table/docs/features/table.md:210]: The beginning of this sentence seems to miss sth.

File: packages/ckeditor5-table/docs/features/table.md [hunk 3]
```diff
@@ -205,6 +205,58 @@ ClassicEditor
...
+	            alignment: 'right',
+	            width: '500px',
+	            height: '250px'
+            }
+        }
+    }
+};
+```
+
+The table element should be aligned to the `right` side by default. Its side should `500x250px`. Border style should be `dashed`, `3px` of its width, and color specified as `“Light green”`. The same touches new tables if they will be inserted into the editor
```
Comment (niegowski) [inline packages/ckeditor5-table/docs/features/table.md:250]:
```suggestion
The table element should be aligned to the `right` side by default. Its size should be `500x250px`. Border style should be `dashed`, `3px` of its width, and color specified as `“Light green”`. The same touches new tables if they will be inserted into the editor
```

Comment (niegowski) [reply packages/ckeditor5-table/docs/features/table.md:250]: > The same touches new tables 

I don't know what you mean here.

**Review by niegowski — CHANGES_REQUESTED** [review]
Partial review - tests not verified.

File: packages/ckeditor5-table/src/tableproperties/tablepropertiesediting.js [hunk 4]
```diff
@@ -69,36 +71,65 @@ export default class TablePropertiesEditing extends Plugin {
 		const schema = editor.model.schema;
 		const conversion = editor.conversion;
 
+		editor.config.define( 'table.tableProperties.defaultProperties', {} );
+
+		const defaultTableProperties = getNormalizedDefaultProperties( editor.config.get( 'table.tableProperties.defaultProperties' ), {
+			includeAlignmentProperty: true
```
[the remark this answers is not in the document]
Comment (pomek) [reply packages/ckeditor5-table/src/tableproperties/tablepropertiesediting.js:77]: Not true. For table cell we have horizontal and vertical aligments.

File: packages/ckeditor5-table/docs/features/table.md (hunk 3 above)
Comment (pomek) [reply packages/ckeditor5-table/docs/features/table.md:250]:
```suggestion
The table element should be aligned to the `right` side by default. Its size should be `500x250px`. Border style should be `dashed`, `3px` of its width, and color specified as `“Light green”`. The same will be applied for new tables if they will be inserted into the editor
```

File: packages/ckeditor5-table/src/tableproperties/tablepropertiesediting.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (niegowski) [reply packages/ckeditor5-table/src/tableproperties/tablepropertiesediting.js:77]: But in this PR there are only default table properties. Is sth missing in this PR?

[the remark this answers is not in the document]
Comment (pomek) [reply packages/ckeditor5-table/src/tableproperties/tablepropertiesediting.js]: Ok, I see what you mean. Fixed.

File: packages/ckeditor5-table/docs/features/table.md (hunk 2 above)
Comment (pomek) [reply packages/ckeditor5-table/docs/features/table.md:210]: Hm?

Comment (niegowski) [reply packages/ckeditor5-table/docs/features/table.md:210]:
```suggestion
The table styles feature allows configuring the default look of tables in the editor. The configuration object should be synchronized with the {@link builds/guides/integration/content-styles editor content styles}.
```

File: packages/ckeditor5-table/tests/tableproperties/tablepropertiesediting.js [hunk 5]
```diff
@@ -1091,6 +1101,13 @@ describe( 'table properties', () => {
 					expect( table.getAttribute( 'alignment' ) ).to.equal( 'left' );
 				} );
 
+				it( 'should discord the unknown float value (style="float:foo;")', () => {
```
Comment (niegowski) [inline packages/ckeditor5-table/tests/tableproperties/tablepropertiesediting.js:1104]:
```suggestion
				it( 'should discard the unknown float value (style="float:foo;")', () => {
```

**Review by niegowski — APPROVED** [review]

**Pull request merged**